pkg/services/ngalert/notifier/alertmanager.go

```diff
@@ -308,6 +309,155 @@ func (am *Alertmanager) SyncAndApplyConfigFromDatabase() error {
+	if err != nil {
+		// The request contains invalid config for one or more contact points
+		return nil, fmt.Errorf("failed to create receivers: %w", err)
+	}
+
+	// job contains all metadata required to test a receiver
+	type job struct {
+		Config       *apimodels.PostableGrafanaReceiver
+		ReceiverName string
+		Notifier     notify.Integration
+	}
+
+	// result contains all metadata required to test a receiver
+	// and an error that is non-nil if the test fails
+	type result struct {
+		Config       *apimodels.PostableGrafanaReceiver
+		ReceiverName string
+		Error        error
+	}
+
+	workers := sync.WaitGroup{}
+	numWorkers := 10
+	workCh := make(chan job, numWorkers)
+	resultCh := make(chan result, numWorkers)
+
+	for i := 0; i < numWorkers; i++ {
+		workers.Add(1)
+		go func() {
+			defer workers.Done()
+			for next := range workCh {
+				v := result{
+					Config:       next.Config,
+					ReceiverName: next.ReceiverName,
+				}
+				if _, err := next.Notifier.Notify(ctx, testAlert); err != nil {
+					v.Error = err
+				}
+				resultCh <- v
+			}
+		}()
+	}
+
+	// start a goroutine that will close the results chan
+	// when all workers have finished
+	go func() {
+		defer close(resultCh)
+		workers.Wait()
+	}()
+
+	// start a goroutine that creates a job per config per receiver
+	go func() {
+		defer close(workCh)
+		for _, receiver := range c.Receivers {
+			for ix := range receiver.GrafanaManagedReceivers {
+				workCh <- job{
+					Config:       receiver.GrafanaManagedReceivers[ix],
+					ReceiverName: receiver.Name,
+					Notifier:     integrationsMap[receiver.Name][ix],
+				}
+			}
+		}
+	}()
+
+	// collate the results from the workers with the correct receivers
+	m := make(map[string]apimodels.TestReceiverResult)
+	for result := range resultCh {
+		tmp := m[result.ReceiverName]
+		tmp.Name = result.ReceiverName
+		status := "ok"
+		if result.Error != nil {
+			status = "failed"
+		}
+		tmp.Configs = append(tmp.Configs, apimodels.TestReceiverConfigResult{
+			Name:   result.Config.Name,
+			Uid:    result.Config.UID,
+			Status: status,
+			Error:  processNotifierError(result.Error),
+		})
+		m[result.ReceiverName] = tmp
+	}
+
+	// prepare the results
+	results := apimodels.TestReceiversResult{
+		Receivers: make([]apimodels.TestReceiverResult, 0, len(c.Receivers)),
+		NotifedAt: now,
+	}
+	for _, next := range m {
+		results.Receivers = append(results.Receivers, next)
+	}
+	return &results, nil
+}
+
+func processNotifierError(err error) string {
+	if err == nil {
+		return ""
+	}
+
+	return err.Error()
+}
+
+func (am *Alertmanager) getTemplate() (*template.Template, error) {
+	am.reloadConfigMtx.RLock()
+	defer am.reloadConfigMtx.RUnlock()
+	if am.config != nil {
+		paths := make([]string, 0, len(am.config.TemplateFiles))
+		for name := range am.config.TemplateFiles {
+			paths = append(paths, filepath.Join(am.WorkingDirPath(), name))
+		}
+		return am.templateFromPaths(paths...)
+	}
+	return nil, errors.New("asd")
```

Comment:
```suggestion
	return nil, errors.New("unable to get templates")
```

Reply:
I think a good error message to return here would be something like:

    return nil, errors.New("alertmanager is not initialized")

Reply:
You're right! Can we make this a guard clause then?

```golang
if am.config != nil {
  return nil, errors.New("alertmanager is not ready yet")
}
```

Reply:
One pattern that I love is to have a `Ready` and `ready`. Such as:

```golang
func (am *Alertmanager) Ready() bool {
	// We consider AM as ready only when the config has been
	// applied at least once successfully. Until then, some objects
	// can still be nil.
	am.reloadConfigMtx.RLock()
	defer am.reloadConfigMtx.RUnlock()
	return am.ready()
}

func (am *Alertmanager) ready() bool {
	return am.config != nil
}
```

Then, we could do here:

```golang
func (am *Alertmanager) getTemplate() (*template.Template, error) {
	am.reloadConfigMtx.RLock()
	defer am.reloadConfigMtx.RUnlock()
	if !am.ready {
	     return nil, errors.New("alertmanager is not ready yet.") 
        }
	
	paths := make([]string, 0, len(am.config.TemplateFiles))
	for name := range am.config.TemplateFiles {
		paths = append(paths, filepath.Join(am.WorkingDirPath(), name))
	}
	return am.templateFromPaths(paths...)
}
```